tests/kernel/sched/schedule_api/src/test_sched_pending.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <ksched.h>
+#include <spinlock.h>
+#include <wait_q.h>
+#include <ztest.h>
+
+struct sched_pending {
+	_wait_q_t waitq;
+	struct k_spinlock lock;
+} pending;
+
+#define STACK_SIZE	(1024 + CONFIG_TEST_EXTRA_STACKSIZE)
+
+K_THREAD_STACK_DEFINE(test_stack, STACK_SIZE);
+
+static struct k_thread tdata;
+static atomic_t num;
+static k_tid_t tid;
+
+static void thread_handler1(void *p1, void *p2, void *p3)
+{
+	int times = (int)p1;
+
+	z_waitq_init(&pending.waitq);
+	for (int i = 0; i < times; i++) {
+		k_spinlock_key_t key = k_spin_lock(&pending.lock);
+
+		z_pend_thread(_current, &pending.waitq, K_FOREVER);
+		z_reschedule(&pending.lock, key);
+	}
+}
+
+static void thread_handler2(void *p1, void *p2, void *p3)
+{
+	z_waitq_init(&pending.waitq);
+
+	unsigned int key = irq_lock();
+
+	z_pend_curr_irqlock(key, &pending.waitq, K_FOREVER);
+}
+
+static void thread_handler3(void *p1, void *p2, void *p3)
+{
+	atomic_inc(&num);
+}
+
+/**
+ * @brief Test kernel api z_pend_thread()  z_unpend_thread()
+ * z_unpend_all() z_pend_curr_irqlock()
+ *
+ * @ingroup kernel_sched_tests
+ *
```

Comment:
I know you have already commented the code. My suggestion is to add @details to descript the high level testing purpose of this case, thanks!